File: src/tinychain/merkle_tree.py
import blake3

class MerkleTree:
    def __init__(self):
        self.leaves = []

    def append(self, data):
        hasher = blake3.blake3()
        hasher.update(data)
        self.leaves.append(hasher.digest())

    def root_hash(self):
        if len(self.leaves) == 0:
            return b''

        while len(self.leaves) > 1:
            new_leaves = []
            for i in range(0, len(self.leaves), 2):
                left_hash = self.leaves[i]
                right_hash = self.leaves[i + 1] if i + 1 < len(self.leaves) else self.leaves[i]
                combined_hash = blake3.blake3(left_hash + right_hash).digest()
                new_leaves.append(combined_hash)
            self.leaves = new_leaves

        return self.leaves[0]

File: src/tinychain-rust/src/merkle_tree.rs
use blake3;

pub struct MerkleTree {
    leaves: Vec<[u8; 32]>,
}

impl MerkleTree {
    pub fn new() -> Self {
        MerkleTree { leaves: Vec::new() }
    }

    pub fn append(&mut self, data: &[u8]) {
        let mut hasher = blake3::Hasher::new();
        hasher.update(data);
        self.leaves.push(hasher.finalize().as_bytes().clone());
    }

    pub fn root_hash(&mut self) -> [u8; 32] {
        if self.leaves.is_empty() {
            return [0u8; 32];
        }

        while self.leaves.len() > 1 {
            let mut new_leaves = Vec::new();
            for i in (0..self.leaves.len()).step_by(2) {
                let left_hash = self.leaves[i];
                let right_hash = if i + 1 < self.leaves.len() {
                    self.leaves[i + 1]
                } else {
                    self.leaves[i]
                };
                let mut hasher = blake3::Hasher::new();
                hasher.update(&left_hash);
                hasher.update(&right_hash);
                new_leaves.push(hasher.finalize().as_bytes().clone());
            }
            self.leaves = new_leaves;
        }

        self.leaves[0]
    }
}
